src/core/qgsnullpainterdevice.h

```diff
@@ -0,0 +1,109 @@
+  Copyright            : (C) 2013 by Mathieu Pellerin
+  Email                : nirvn dot asia at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef QGSNULLPAINTERDEVICE_H
+#define QGSNULLPAINTERDEVICE_H
+
+#define SIP_NO_FILE
+
+#include "qgis_core.h"
+
+#include <QPaintDevice>
+#include <QPaintEngine>
+
+
+class QgsNullPaintEngine: public QPaintEngine
+{
+
+  public:
+
+    QgsNullPaintEngine() : QPaintEngine( QPaintEngine::AllFeatures ) {};
+
+    bool begin( QPaintDevice * ) override { return true; };
+    bool end() override { return true; };
+    QPaintEngine::Type type() const override { return QPaintEngine::User; };
+    void updateState( const QPaintEngineState & ) override { return; };
+    void drawPixmap( const QRectF &, const QPixmap &, const QRectF & ) override { return; };
+
+};
+
+
+/**
+ * \ingroup core
+ * \brief Null painter device that can be used for map renderer jobs which use custom painters.
+ *
+ * \since QGIS 3.24
+ */
+class CORE_EXPORT QgsNullPaintDevice: public QPaintDevice
+{
+
+  public:
+
+    QgsNullPaintDevice()
+    {
+      mPaintEngine = std::make_unique<QgsNullPaintEngine>();
+    };
+
+    QPaintEngine *paintEngine() const override { return mPaintEngine.get(); };
+    int metric( PaintDeviceMetric metric ) const override
+    {
+      switch ( metric )
+      {
+        case QPaintDevice::PdmWidth:
+          return mSize.width();
+        case QPaintDevice::PdmHeight:
+          return mSize.height();
+        case QPaintDevice::PdmWidthMM:
+          return mSize.width();
+        case QPaintDevice::PdmHeightMM:
+          return mSize.height();
+        case QPaintDevice::PdmNumColors:
+          return std::numeric_limits<int>::max();
+        case QPaintDevice::PdmDepth:
+          return 32;
+        case QPaintDevice::PdmDpiX:
+        case QPaintDevice::PdmDpiY:
+        case QPaintDevice::PdmPhysicalDpiX:
+        case QPaintDevice::PdmPhysicalDpiY:
+          return mDpi;
+        case QPaintDevice::PdmDevicePixelRatio:
+          return 1;
+        case QPaintDevice::PdmDevicePixelRatioScaled:
+          return 1;
+      }
+      return 0;
+    };
+
+    /**
+     * Sets the \a size of the device in pixels.
+     */
+    void setOutputSize( const QSize &size ) { mSize = size; };
+
+    /**
+     * Sets the \a dpi of the device.
+     */
+    void setOutputDpi( const int dpi ) { mDpi = dpi; };
```

Comment:
This should be double

Reply:
@nyalldawson , should it here? That value is used to return logicalDpi{X,Y} through the overridden metric class function. That function returns ints.